examples/friend-list/src/application/saveQueryToHistory.ts

```diff
@@ -0,0 +1,45 @@
+import { hold, map, merge, skipRepeats, startWith, tap } from '@motorcycle/stream'
+import { pipe, prop, reduce, set } from '167'
+
+import { Stream } from '@motorcycle/types'
+import { popState } from './popState'
+
+const QUERY_PARAM_NAME = 'q'
+
+const saveQuery = pipe(skipRepeats, tap(pushQuery))
+const retrieveStateOnPopstate = merge(map(retrieveCurrentState, popState()))
+const startWithInitialState = startWith(retrieveCurrentState())
+
+export const saveQueryToHistory = pipe<Stream<string>, Stream<string>>(
+  saveQuery,
+  retrieveStateOnPopstate,
+  startWithInitialState,
+  skipRepeats,
+  hold
+)
+
+function pushQuery(value: string) {
+  const pathName = value.trim() ? `/?${QUERY_PARAM_NAME}=${value.trim()}` : '/'
+
+  history.pushState(null, null, pathName)
```

Comment:
I don’t think it is correct to tie the application to the browser. Instead, like we use `DocumentDomSource` in the sinks, so should `history` and `location` be provided to the UI, which pass values to the application.